backend/request.go

```diff
@@ -86,6 +87,55 @@ func GetHost(r *http.Request) string {
 	return "https://" + r.Host
 }
 
+func stringMap(vs []string, f func(string) string) []string {
+	vsm := make([]string, len(vs))
+	for i, v := range vs {
+		vsm[i] = f(v)
+	}
+	return vsm
+}
+
+// Converts header entries associated with key to canonical form. In particular,
+// multiple headers are collapsed into one.
+func canonical(h http.Header, key string) {
+	v := h[textproto.CanonicalMIMEHeaderKey(key)]
+	var a []string
+	for _, vv := range v {
+		if vv != "" {
+			a = append(a, stringMap(strings.Split(vv, ","), strings.TrimSpace)...)
+		}
+	}
+	if len(a) != 0 {
+		h[textproto.CanonicalMIMEHeaderKey(key)] = []string{strings.Join(a, ", ")}
+	}
+}
+
+// Adds value associated with header if not already present. Assumes keys are
+// unique.
+func add(h http.Header, key string, value string) {
+	v := h[textproto.CanonicalMIMEHeaderKey(key)]
+	if len(v) == 0 {
+		h[textproto.CanonicalMIMEHeaderKey(key)] = []string{value}
+	} else {
+		a := stringMap(strings.Split(v[0], ","), strings.TrimSpace)
+		for _, vv := range a {
+			if vv == value {
+				return
+			}
+		}
+		h[textproto.CanonicalMIMEHeaderKey(key)] = []string{strings.Join(append(a, value), ", ")}
+	}
+}
+
+func SetVary(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		canonical(w.Header(), "vary")
+		add(w.Header(), "vary", "Accept")
+		add(w.Header(), "vary", "AMP-Cache-Transform")
```

Comment:
Well … partly! I was enjoying the go HTTP libraries so much that I thought it would be super straightforward to dedupe headers, but it turns out that whilst pretty much everything you might want to do re the networking side of HTTP is possible via the standard libraries the header munging part is not. Hence, this hassle…

I don't have any particular reason to believe this is actually necessary, but you never really know when it comes to HTTP…